### ISSUE TITLE
make machines pickleable

### ISSUE THREAD

**Issue by ajax2leet** [tyarkoni/transitions#119]
Currently, machines are not pickleable, due to the instancemethods added to the model. In principle, this is easy to circumvent by removing the instancemethods in the __getstate__ method of the model class. Upon recreating the object from the pickle file, the instancemethods would have to be again added to the model.
To make this work, the setattr calls in add_states and add_transitions would have to be moved to sperarate helper functions, so that they can be called by __setstate__ upon reload the pickled object.

**Comment by aleneum** [comment]
Do you work with python 2.7 or python 3.3 or older ?
If thats so, you can use dill to pickle machine instances. Have a look at the tests in `test/test_core.py` or `test/test_nested.py`:

For instance [here](https://github.com/tyarkoni/transitions/blob/master/tests/test_nesting.py#L254):

```python
def test_pickle(self):
        import sys
        if sys.version_info < (3, 4):
            import dill as pickle
        else:
            import pickle

        states = ['A', 'B', {'name': 'C', 'children': ['1', '2', {'name': '3', 'children': ['a', 'b', 'c']}]},
          'D', 'E', 'F']
        transitions = [
            {'trigger': 'walk', 'source': 'A', 'dest': 'B'},
            {'trigger': 'run', 'source': 'B', 'dest': 'C'},
            {'trigger': 'sprint', 'source': 'C', 'dest': 'D'}
        ]
        m = self.stuff.machine_cls(states=states, transitions=transitions, initial='A')
        m.walk()
        dump = pickle.dumps(m)
        self.assertIsNotNone(dump)
        m2 = pickle.loads(dump)
        self.assertEqual(m.state, m2.state)
        m2.run()
        if State.separator in '_':
            m2.to_C_3_a()
            m2.to_C_3_b()
        else:
            m2.to_C.s3.a()
            m2.to_C.s3.b()
```

**Comment by ajax2leet** [comment]
Ah, thats good to know. I'm in fact using python 2.7.3. But dill works fine, thanks!

**Comment by ajax2leet** [comment]
Whops, actually it doesn't work. There seems to be a problem with conditions:

```python
from transitions import Machine
import dill 

class Model:
    def condition(self):
        return False

states = ['A', 'B', 'C', 'D']
transitions = [
    {'trigger': 'walk', 'source': 'A', 'dest': 'B','conditions':'condition'},
    {'trigger': 'run', 'source': 'B', 'dest': 'C'},
    {'trigger': 'sprint', 'source': 'C', 'dest': 'D'}
]

m = Model()

ms = Machine(m,states=states, transitions=transitions, initial='A')
m.walk()
dump = dill.dumps(ms)       
ms2 = dill.loads(dump)
ms2.run()
```
The example raises a PicklingError:

```
PicklingError: Can't pickle <class 'transitions.core.Condition'>: it's not found as transitions.core.Condition
```

**Comment by aleneum** [comment]
Running the nose tests does not throw any errors but I do have the same issue when executing a test from an external file 😕. I am digging my way through several dill issues ([[1]](https://github.com/ipython/ipyparallel/issues/113)[[2]](https://github.com/uqfoundation/dill/issues/135)[[3]](https://github.com/uqfoundation/dill/issues/56)) and it seems to be related with how dill decides what object requires special treatment and what not.
Maybe there is a way to configure dill so that it 'detects' Condition right because it really should be able to do so.

Some people also mention `cloudpickle` which does not have this specific issue. I replaced dill with cloudpickle and things seem to work. Could you give cloudpickle a try if it works for your scenario?

**Comment by tyarkoni** [comment]
Is this an issue with Condition being nested inside Transition? If so, we could simply move Condition out of there. I'd prefer to keep it encapsulated since nothing else uses it, but if that's what's tripping up dill, this would be an easy fix.

**Comment by aleneum** [comment]
That would work, yeah. But the encapsulation makes sense. And dill can handle nesting. I am positive that sooner or later dill we be capable of dealing with this specific issue since it seems to be a known issue. For the moment, I am looking into 'workarounds' other projects have tried so far. If there is no clean solution, I am all for the easy fix.

**Comment by ajax2leet** [comment]
Unfortunately Cloudpickle doesn't work. I haven't tracked down why, but for my usecase it fails with an AssertionError:
```
../lib/python2.7/pickle.pyc in memoize(self, obj)
242         if self.fast:
    243             return
--> 244         assert id(obj) not in self.memo
    245         memo_len = len(self.memo)
    246         self.write(self.put(memo_len))
AssertionError: 
```

**Comment by aleneum** [comment]
bummer. thanks for the try.

**Comment by aleneum** [comment]
After some digging it seems that @tyarkoni's is by far the most efficient solution. 👍

**Comment by tyarkoni** [comment]
Fixed in #120. Hopefully it works now, @ajax2leet!
